### ISSUE TITLE
Worker thread panics (cudarc `dst.len() >= src.len()`) under high concurrency → permanent engine wedge while /health stays 200

### ISSUE THREAD

**Issue by xiaguan** [xiaguan/pegainfer#403]
## Summary

Under sustained high concurrency (mixed prefill + decode), the Qwen3-4B GPU worker thread (`qwen3-tp-rank-0`) panics inside a cudarc memory copy:

```
thread 'qwen3-tp-rank-0' panicked at cudarc-0.19.7/src/driver/safe/core.rs:1607:9:
assertion failed: dst.len() >= src.len()
```

A pre-allocated buffer is too small for some batch/shape a step produces. The panic kills the worker thread, and the engine **never recovers**: the scheduler then fails the entire in-flight batch and every subsequent request with `tensor-parallel worker step channel closed`, while `/health` keeps returning `200`. The result is a **silent, permanent wedge** that only a process restart clears.

This is **not** a KV-admission / over-subscription problem (see "What it is NOT" below).

## Severity

Two distinct faults compound:

1. **Undersized buffer** — the proximate crash (`dst.len() >= src.len()` in a cudarc copy).
2. **Blast radius + no recovery** — when any step returns `Err`, `fail_touched_requests` does `active.clear()`, failing *every* in-flight request (not just the one with the bad shape). The worker thread is never restarted, so all subsequent steps return `worker step channel closed` and **all future requests fail too**. `/health` is a separate layer and still answers `200`, so the dead worker is invisible to liveness checks.

Verified after the crash: `ps -T` shows no `qwen3-tp-rank-0` thread, and a trivial `2+2=` completion fails — the engine is dead until the process is restarted.

## Reproduction

Hardware note: easiest to hit on a card with a small KV pool. On a 16 GB RTX 5070 Ti, after the 7.67 GB model load only ~6.5 GB is free → KV pool = 2473 blocks (~39.5K tokens). The tight pool forces mixed prefill+decode steps under load, which is where the bad shape appears. On larger cards, push `--concurrency` higher to reach the same regime.

1. Build and start the server (default `qwen3-4b` feature):
   ```bash
   cargo run --release -- --model-path models/Qwen3-4B
   ```
2. Drive sustained high-concurrency, decode-heavy load with the in-repo bench:
   ```bash
   python scripts/bench_http_serving.py \
     --model models/Qwen3-4B \
     --concurrency 128 --num-requests 384 \
     --prompt-words 32 --max-tokens 256 --temperature 0
   ```
3. Within a few runs (it is intermittent — it depends on hitting a particular batch composition), the worker panics. Symptoms:
   - the bench reports most/all requests `failed`;
   - the server log shows the `qwen3-tp-rank-0` panic followed by a flood of `Execution step failed: tensor-parallel worker step channel closed`;
   - every request after the panic fails, yet `curl http://127.0.0.1:8000/health` still returns `200`.

A plain concurrent burst of streaming `/v1/completions` requests (≈128–192 in flight, 32-word prompts, `max_tokens=256`) reproduces it equally well without the bench script.

## Observed logs

```
thread 'qwen3-tp-rank-0' (518213) panicked at cudarc-0.19.7/src/driver/safe/core.rs:1607:9:
assertion failed: dst.len() >= src.len()
...
WARN openinfer_qwen3_4b::scheduler: scheduler.rs:451 Execution step failed: primary worker dropped step response
WARN openinfer_qwen3_4b::scheduler: scheduler.rs:451 Execution step failed: tensor-parallel worker step channel closed   (×206)
ERROR vllm_engine_core_client::client::stream: stream.rs:91 request failed with an internal error during generation
```

## What it is NOT

- **Not KV admission / over-subscription.** A single one-shot burst of 160 concurrent requests (`max_tokens=256`) completes with **0 errors** — full-lifetime admission (`request_lifetime_blocks` reserving `prompt + max_tokens`) correctly *defers* the overflow. The failure only appears under sustained mixed prefill+decode load, and it is a worker panic, not a `Rejected`/admission event.
- The client-facing error is a generic `{"message":"Internal server error",...}`; the real cause is only visible in the server log (the panic + `channel closed`).

## Where to look / next

- The panic is a cudarc copy with `dst` smaller than `src`. The decode token H2D (`batch_decode.rs`, `memcpy_htod(token_ids_padded → token_ids_d)`) is sized to the 256-wide bucket and `padded_bs ≤ 256`, so it is **not** the suspect.
- Prime suspects are the **prefill / unified step** or the **batched step-tail logits/sampling readback** paths, where a host or device buffer may be sized from something other than the actual per-step row/token count. Run with `RUST_BACKTRACE=1` to capture the exact call site.
- Independently worth fixing: the blast radius. A single step error should fail (or preempt) only the offending request, and a dead worker should either be restarted or surfaced through `/health`, instead of wedging the whole engine silently.

## Environment

- Commit: `0d3af7c` (main)
- Model: Qwen3-4B (default `qwen3-4b` feature, single GPU, `tp_size=1`)
- GPU: NVIDIA GeForce RTX 5070 Ti (16 GB), driver 580.126.09
- CUDA: 13.3
- cudarc: 0.19.7

**Comment by n-WN** [comment]
Root-caused fault 1. The undersized buffer is `BatchDecodeBuffers::page_indices_d`, and the trigger is **prefix-cache block sharing**, not batch composition:

- `page_indices_d` is sized `total_blocks + max_bs` — a *physical pool* bound.
- Prefix-matched blocks are shared `ImmutableBlock`s, so N views holding the same cached prefix each list those page ids again in `sync_paged_meta`'s concatenated list. The list is counted **by reference**; with enough sharing it exceeds any physical-pool bound.
- Admission can't save it: `request_lifetime_blocks` charges physical blocks, and sharing keeps physical usage far below the by-reference count.

Backtrace (H100, main @ `1275262`, `RUST_BACKTRACE=1`):

```
3: cudarc::driver::safe::core::CudaStream::memcpy_htod::<i32, ...>
4: openinfer_qwen3_4b::batch_decode_buffers::BatchDecodeBuffers::sync_paged_meta
5: openinfer_qwen3_4b::weights::Qwen3Model::batch_decode
6: openinfer_qwen3_4b::executor::execute_step_on_lane
```

This explains both puzzles in the report:

- **Why the one-shot 160 burst is clean**: blocks only become matchable once sealed, and simultaneous prefills don't match each other. Sharing needs sustained arrivals over an already-sealed prefix — exactly "sustained mixed prefill+decode load".
- **Why it's intermittent on the 5070 Ti**: `scripts/bench_http_serving.py`'s `DEFAULT_PROMPT_WORDS` has 24 words, so `--prompt-words 32` produces only 24 distinct prompts. The shared double-count across in-flight requests hovers right at the `+ max_bs` slack — whether a given run crosses it depends on which duplicates are co-resident.

With an explicitly shared prefix it is deterministic, and big cards make it *easier*, not harder (more concurrent shared views). One-file repro against a default `qwen3-4b` server:

```python
# warm one long prompt, then fire N identical requests -> every decode view
# lists the ~same shared pages; sum(views) >> pool + max_bs.
import json, threading, urllib.request
BASE, MODEL = "http://127.0.0.1:8000", "<model-path>"
def go(prompt, max_tokens):
    body = json.dumps({"model": MODEL, "prompt": prompt, "max_tokens": max_tokens,
                       "temperature": 0.0}).encode()
    req = urllib.request.Request(BASE + "/v1/completions", data=body,
                                 headers={"Content-Type": "application/json"})
    try:
        with urllib.request.urlopen(req, timeout=300) as r: return r.status
    except Exception as e: return getattr(e, "code", repr(e)[:60])
prompt = " ".join(f"item{i%97} value{i%89}" for i in range(3000))
print("warm:", go(prompt, 1))
results = []
ts = [threading.Thread(target=lambda: results.append(go(prompt, 64))) for _ in range(256)]
[t.start() for t in ts]; [t.join() for t in ts]
from collections import Counter; print(Counter(results))
print("post:", go("2+2=", 4))  # wedged engine -> 500 while /health stays 200
```

On this box (H100 80GB, pool 24201 blocks): warm OK → 63/256 requests 500, the panic above, and the permanent wedge (`2+2=` → 500, `/health` → 200). PR incoming with the sizing fix (by-reference bound: `max_batch × full-context view`), a fail-loud guard in `sync_paged_meta`, and a GPU regression test; verified 256/256 clean on the same repro after the fix.
